File: components/mutations/UncancelTweet.ts
import { commitMutation,Environment, graphql } from "react-relay"

import { UncancelTweetMutationResponse } from "@generated/UncancelTweetMutation.graphql"

const mutation = graphql`
  mutation UncancelTweetMutation($input: UncancelTweetInput!) {
    uncancelTweet(input: $input) {
      tweet {
        id
        status
      }
    }
  }
`

function getOptimisticResponse(id: string): UncancelTweetMutationResponse {
  return {
    uncancelTweet: {
      tweet: {
        id,
        status: "DRAFT",
      },
    },
  }
}

export function uncancelTweet(environment: Environment, id: string): void {
  const variables = { input: { id } }

  commitMutation(environment, {
    mutation,
    variables,
    optimisticResponse: getOptimisticResponse(id),
    // TODO add update config to move this to the other range
  })
}

File: ios/Courier/Courier/Mutations/UncancelTweet.swift
import Relay
import RelaySwiftUI
import CourierGenerated

private let mutation = graphql("""
mutation UncancelTweetMutation($input: UncancelTweetInput!) {
  uncancelTweet(input: $input) {
    tweetGroup {
      id
      status
    }
  }
}
""")

private func updater(_ store: inout RecordSourceSelectorProxy, _ data: SelectorData?) {
    guard let selectorData = data else {
        return
    }

    let data = try! SelectorDataDecoder().decode(UncancelTweetMutation.Data.self, from: selectorData)
    moveTweetGroup(&store, id: data.uncancelTweet.tweetGroup.id, from: .past, to: .upcoming)
}

extension Mutation.Mutator where Operation == UncancelTweetMutation {
    func commit(id: String) {
        commit(
            variables: .init(input: UncancelTweetInput(id: id)),
            optimisticResponse: [
                "uncancelTweet": [
                    "tweetGroup": [
                        "id": id,
                        "status": "DRAFT",
                    ]
                ]
            ],
            optimisticUpdater: updater,
            updater: updater
        )
    }
}
